sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -155,8 +156,9 @@ private[sql] class CacheManager extends Logging {
    * function will over invalidate.
    */
   private[sql] def invalidateCache(plan: LogicalPlan): Unit = writeLock {
+    val canonicalized = plan.canonicalized
     cachedData.foreach {
-      case data if data.plan.collect { case p if p.sameResult(plan) => p }.nonEmpty =>
+      case data if data.plan.collect { case p if p.sameResult(canonicalized) => p }.nonEmpty =>
```

Comment:
Some plans like `LocalRelation`, `LogicalRDD`, etc. override the `sameResult` and it does not use canonicalized plan to compare.

For example, `LocalRelation` overrides like as follows:

```
  override def sameResult(plan: LogicalPlan): Boolean = plan match {
    case LocalRelation(otherOutput, otherData) =>
      otherOutput.map(_.dataType) == output.map(_.dataType) && otherData == data
    case _ => false
  }
```

If the `plan` is `SubqueryAlias(LocalRelation)`, the method returns `false`.